### PR TITLE
[City-gates] Determine and create city gates

### PR DESCRIPTION
**Pull request by NicEastvillage**
Adds functionality for the script to determine city gates and write those gates to the statistics file. The number of gates can be set by the user (default is 4). If there are already gates defined in the stats file, only missing gates are added.

The gates are placed by creating N arrows pointing in different directions. The dead end nodes furthest in those directions are chosen as city gates. The amount of traffic going through the gates are determined by ActivityGen, so we just have to decide the proportion. For now I made it uniformly random in the interval [1;2).

This method might not work in all cases. Edge cases are to be found and eliminated.

### CODE REVIEW

File: randomActivityGen.py [hunk 1]
```diff
@@ -28,15 +33,74 @@
...
+
+def setup_city_gates(net: sumolib.net.Net, stats: ET.ElementTree, gate_count: int):
+    assert gate_count >= 0, "Number of city gates cannot be negative"
+    # Find existing gates to determine how many we need to insert
+    xml_gates = stats.find("cityGates")
+    xml_entrances = xml_gates.findall("entrance")
+    n = gate_count - len(xml_entrances)
+    if n < 0:
+        print(f"Warning: {gate_count} city gate was requested, but there are already {len(xml_entrances)} defined")
+    if n <= 0:
```
Comment (MathiasMehl) [inline randomActivityGen.py:53]: Would it not make more sense that it is the if statement on line 51 that is n <= 0? 
Does not make sense to insert 0 new city gates

**Review by MathiasMehl — COMMENTED** [review]
Looks good otherwise, math checks out

File: randomActivityGen.py (hunk 1 above)
Comment (NicEastvillage) [reply randomActivityGen.py:53]: If 0 gates needs to be inserted the function returns on line 54. It is not unreasonable to want 0 gates in the city. But if the user wants less gates than already defined in the stats file then we will warn the user (and also return on line 54).

**Review by MathiasMehl — APPROVED** [review]

File: randomActivityGen.py (hunk 1 above)
Comment (MathiasMehl) [reply randomActivityGen.py:53]: I see, I guess it confused me that line 53 says n <= 0, when its simply a check if n == 0?

Comment (NicEastvillage) [reply randomActivityGen.py:53]: But it's not. We warn on negative and abort on negative and zero.
Anyway, it's merged.

**Pull request merged**